File: packages/liveblocks-node/src/webhooks.ts
import type { NotificationChannel } from "@liveblocks/core";
import * as base64 from "@stablelib/base64";
import * as sha256 from "fast-sha256";
import type { IncomingHttpHeaders } from "http";

import { isString } from "./utils";

export class WebhookHandler {
  #secretBuffer: Buffer;
  static #secretPrefix = "whsec_";

  constructor(
    /**
     * The signing secret provided on the dashboard's webhooks page
     * @example "whsec_wPbvQ+u3VtN2e2tRPDKchQ1tBZ3svaHLm"
     */
    secret: string
  ) {
    if (!secret) throw new Error("Secret is required");
    if (typeof secret !== "string") throw new Error("Secret must be a string");

    if (secret.startsWith(WebhookHandler.#secretPrefix) === false)
      throw new Error("Invalid secret, must start with whsec_");

    const secretKey = secret.slice(WebhookHandler.#secretPrefix.length);
    this.#secretBuffer = Buffer.from(secretKey, "base64");
  }

  /**
   * Verifies a webhook request and returns the event
   */
  public verifyRequest(request: WebhookRequest): WebhookEvent {
    const { headers, rawBody } = request;

    const { webhookId, timestamp, rawSignatures } =
      this.#verifyHeaders(headers);

    if (typeof rawBody !== "string") {
      throw new Error(
        `Invalid rawBody field, must be a string, got "${typeof rawBody}" instead. It is likely that you need to JSON.stringify the body before passing it.`
      );
    }

    this.#verifyTimestamp(timestamp);

    const signature = this.#sign(`${webhookId}.${timestamp}.${rawBody}`);

    const expectedSignatures = rawSignatures
      .split(" ")
      .map((rawSignature) => {
        const [, parsedSignature] = rawSignature.split(",");
        return parsedSignature;
      })
      .filter(isNotUndefined);

    if (expectedSignatures.includes(signature) === false)
      throw new Error(
        `Invalid signature, expected one of ${expectedSignatures.join(
          ", "
        )}, got ${signature}`
      );

    const event: WebhookEvent = JSON.parse(rawBody) as WebhookEvent;

    this.#verifyWebhookEventType(event);

    return event;
  }

  /**
   * Verifies the headers and returns the webhookId, timestamp and rawSignatures
   */
  #verifyHeaders(headers: IncomingHttpHeaders | Headers) {
    const usingNativeHeaders =
      typeof Headers !== "undefined" && headers instanceof Headers;
    const normalizedHeaders = usingNativeHeaders
      ? Object.fromEntries(headers)
      : (headers as IncomingHttpHeaders);

    const sanitizedHeaders: IncomingHttpHeaders = {};
    Object.keys(normalizedHeaders).forEach((key) => {
      sanitizedHeaders[key.toLowerCase()] = normalizedHeaders[key];
    });

    const webhookId = sanitizedHeaders["webhook-id"];
    if (typeof webhookId !== "string")
      throw new Error("Invalid webhook-id header");

    const timestamp = sanitizedHeaders["webhook-timestamp"];
    if (typeof timestamp !== "string")
      throw new Error("Invalid webhook-timestamp header");

    const rawSignatures = sanitizedHeaders["webhook-signature"];
    if (typeof rawSignatures !== "string")
      throw new Error("Invalid webhook-signature header");

    return { webhookId, timestamp, rawSignatures };
  }

  /**
   * Signs the content with the secret
   * @param content
   * @returns `string`
   */
  #sign(content: string): string {
    const encoder = new TextEncoder();
    const toSign = encoder.encode(content);
    return base64.encode(sha256.hmac(this.#secretBuffer, toSign));
  }

  /**
   * Verifies that the timestamp is not too old or in the future
   */
  #verifyTimestamp(timestampHeader: string) {
    const now = Math.floor(Date.now() / 1000);
    const timestamp = parseInt(timestampHeader, 10);

    if (isNaN(timestamp)) {
      throw new Error("Invalid timestamp");
    }

    // Check if timestamp is too old
    if (timestamp < now - WEBHOOK_TOLERANCE_IN_SECONDS) {
      throw new Error("Timestamp too old");
    }

    // Check if timestamp is in the future
    if (timestamp > now + WEBHOOK_TOLERANCE_IN_SECONDS) {
      throw new Error("Timestamp in the future");
    }
  }

  /**
   * Ensures that the event is a known event type
   * or throws and prompts the user to upgrade to a higher version of @liveblocks/node
   */
  #verifyWebhookEventType(event: WebhookEvent): asserts event is WebhookEvent {
    if (
      event &&
      event.type &&
      [
        "storageUpdated",
        "userEntered",
        "userLeft",
        "roomCreated",
        "roomDeleted",
        "commentCreated",
        "commentEdited",
        "commentDeleted",
        "commentReactionAdded",
        "commentReactionRemoved",
        "commentMetadataUpdated",
        "threadMetadataUpdated",
        "threadCreated",
        "threadDeleted",
        "ydocUpdated",
        "notification",
        "threadMarkedAsResolved",
        "threadMarkedAsUnresolved",
      ].includes(event.type)
    ) {
      if (event.type === "notification") {
        const notification = event;
        if (
          notification.data.kind === "thread" ||
          notification.data.kind === "textMention" ||
          isCustomKind(notification.data.kind)
        ) {
          return;
        } else {
          // Using JSON.stringify because `notification.data.kind`
          // is considered as `never` now because of the type guard.
          throw new Error(
            `Unknown notification kind: ${JSON.stringify(notification.data.kind)}`
          );
        }
      }

      return;
    }

    throw new Error(
      "Unknown event type, please upgrade to a higher version of @liveblocks/node"
    );
  }
}

const WEBHOOK_TOLERANCE_IN_SECONDS = 5 * 60; // 5 minutes

const isNotUndefined = <T>(value: T | undefined): value is T =>
  value !== undefined;

type WebhookRequest = {
  /**
   * Headers of the request, can be a regular object or a Headers object
   * @example
   * {
   *  "webhook-id": "123",
   *  "webhook-timestamp": "1614588800000",
   *  "webhook-signature": "v1,bm9ldHUjKzFob2VudXRob2VodWUzMjRvdWVvdW9ldQo= v2,MzJsNDk4MzI0K2VvdSMjMTEjQEBAQDEyMzMzMzEyMwo="
   * }
   *
   * new Headers({
   *  "webhook-id": "123",
   *  "webhook-timestamp": "1614588800000",
   *  "webhook-signature": "v1,bm9ldHUjKzFob2VudXRob2VodWUzMjRvdWVvdW9ldQo= v2,MzJsNDk4MzI0K2VvdSMjMTEjQEBAQDEyMzMzMzEyMwo="
   * }}
   */
  headers: IncomingHttpHeaders | Headers;
  /**
   * Raw body of the request, do not parse it
   * @example '{"type":"storageUpdated","data":{"roomId":"my-room-id","appId":"my-app-id","updatedAt":"2021-03-01T12:00:00.000Z"}}'
   */
  rawBody: string;
};

/**
 * When receiving an event we cannot define the `kind`
 * as member of the augmentation
 */
type CustomKind = `$${string}`;

const isCustomKind = (value: unknown): value is CustomKind => {
  return isString(value) && value.startsWith("$");
};

type WebhookEvent =
  | StorageUpdatedEvent
  | UserEnteredEvent
  | UserLeftEvent
  | RoomCreatedEvent
  | RoomDeletedEvent
  | CommentCreatedEvent
  | CommentEditedEvent
  | CommentDeletedEvent
  | CommentReactionAdded
  | CommentReactionRemoved
  | CommentMetadataUpdatedEvent
  | ThreadMetadataUpdatedEvent
  | NotificationEvent
  | ThreadCreatedEvent
  | ThreadDeletedEvent
  | ThreadMarkedAsResolvedEvent
  | ThreadMarkedAsUnresolvedEvent
  | YDocUpdatedEvent;

type StorageUpdatedEvent = {
  type: "storageUpdated";
  data: {
    roomId: string;
    projectId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    updatedAt: string;
  };
};

type UserEnteredEvent = {
  type: "userEntered";
  data: {
    projectId: string;
    roomId: string;
    connectionId: number;
    userId: string | null;
    userInfo: Record<string, unknown> | null;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     * @description The time when the user entered the room.
     */
    enteredAt: string;
    numActiveUsers: number;
  };
};

type UserLeftEvent = {
  type: "userLeft";
  data: {
    projectId: string;
    roomId: string;
    connectionId: number;
    userId: string | null;
    userInfo: Record<string, unknown> | null;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     * @description The time when the user left the room.
     */
    leftAt: string;
    numActiveUsers: number;
  };
};

type RoomCreatedEvent = {
  type: "roomCreated";
  data: {
    projectId: string;
    roomId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    createdAt: string;
  };
};

type RoomDeletedEvent = {
  type: "roomDeleted";
  data: {
    projectId: string;
    roomId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    deletedAt: string;
  };
};

type CommentCreatedEvent = {
  type: "commentCreated";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    commentId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    createdAt: string;
    createdBy: string;
  };
};

type CommentEditedEvent = {
  type: "commentEdited";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    commentId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    editedAt: string;
  };
};

type CommentDeletedEvent = {
  type: "commentDeleted";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    commentId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    deletedAt: string;
  };
};

type CommentReactionAdded = {
  type: "commentReactionAdded";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    commentId: string;
    emoji: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    addedAt: string;
    addedBy: string;
  };
};

type CommentReactionRemoved = {
  type: "commentReactionRemoved";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    commentId: string;
    emoji: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    removedAt: string;
    removedBy: string;
  };
};

type YDocUpdatedEvent = {
  type: "ydocUpdated";
  data: {
    projectId: string;
    roomId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    updatedAt: string;
  };
};

type ThreadMetadataUpdatedEvent = {
  type: "threadMetadataUpdated";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    updatedAt: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    updatedBy: string;
  };
};

type CommentMetadataUpdatedEvent = {
  type: "commentMetadataUpdated";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    commentId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    updatedAt: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    updatedBy: string;
  };
};

type ThreadCreatedEvent = {
  type: "threadCreated";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    createdAt: string;
    createdBy: string;
  };
};

type ThreadDeletedEvent = {
  type: "threadDeleted";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    deletedAt: string;
  };
};

type ThreadMarkedAsResolvedEvent = {
  type: "threadMarkedAsResolved";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    updatedAt: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    updatedBy: string;
  };
};

type ThreadMarkedAsUnresolvedEvent = {
  type: "threadMarkedAsUnresolved";
  data: {
    projectId: string;
    roomId: string;
    threadId: string;
    updatedAt: string;
    /**
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    updatedBy: string;
  };
};

type ThreadNotificationEvent = {
  type: "notification";
  data: {
    channel: NotificationChannel;
    kind: "thread";
    projectId: string;
    roomId: string;
    userId: string;
    threadId: string;
    inboxNotificationId: string;
    /**
     * Date representing the time when the webhook event was created.
     *
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    createdAt: string;
    /**
     * Date representing the  time when the notification was created.
     *
     * ISO 8601 datestring
     * @example "2021-03-01T18:00:00.000Z"
     */
    triggeredAt: string;
  };
};

type TextMentionNotificationEvent = {
  type: "notification";
  data: {
    channel: NotificationChannel;
    kind: "textMention";
    projectId: string;
    roomId: string;
    userId: string;
    mentionId: string;
    inboxNotificationId: string;
    /**
     * Date representing the time when the webhook event was created.
     *
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    createdAt: string;
    /**
     * Date representing the  time when the notification was created.
     *
     * ISO 8601 datestring
     * @example "2021-03-01T18:00:00.000Z"
     */
    triggeredAt: string;
  };
};

type CustomNotificationEvent = {
  type: "notification";
  data: {
    channel: NotificationChannel;
    kind: CustomKind;
    projectId: string;
    roomId: string | null;
    userId: string;
    subjectId: string;
    inboxNotificationId: string;
    /**
     * Date representing the time when the webhook event was created.
     *
     * ISO 8601 datestring
     * @example "2021-03-01T12:00:00.000Z"
     */
    createdAt: string;
    /**
     * Date representing the  time when the notification was created.
     *
     * ISO 8601 datestring
     * @example "2021-03-01T18:00:00.000Z"
     */
    triggeredAt: string;
  };
};

type NotificationEvent =
  | ThreadNotificationEvent
  | TextMentionNotificationEvent
  | CustomNotificationEvent;

export type {
  CommentCreatedEvent,
  CommentDeletedEvent,
  CommentEditedEvent,
  CommentMetadataUpdatedEvent,
  CommentReactionAdded,
  CommentReactionRemoved,
  CustomNotificationEvent,
  NotificationEvent,
  RoomCreatedEvent,
  RoomDeletedEvent,
  StorageUpdatedEvent,
  TextMentionNotificationEvent,
  ThreadCreatedEvent,
  ThreadDeletedEvent,
  ThreadMarkedAsResolvedEvent,
  ThreadMarkedAsUnresolvedEvent,
  ThreadMetadataUpdatedEvent,
  ThreadNotificationEvent,
  UserEnteredEvent,
  UserLeftEvent,
  WebhookEvent,
  WebhookRequest,
  YDocUpdatedEvent,
};

/**
 * Type guard to check if a webhook event is a `ThreadNotificationEvent`
 *
 * The check is made against the event type and event data kind.
 * You should use this guard to safely check the webhook event you received
 * when you're expecting a `ThreadNotificationEvent`.
 *
 * @param event The webhook event received after calling `webhookHandler.verifyRequest()`.
 * @returns A boolean type predicate.
 */
export function isThreadNotificationEvent(
  event: WebhookEvent
): event is ThreadNotificationEvent {
  return event.type === "notification" && event.data.kind === "thread";
}

/**
 * Type guard to check if a webhook event is a `TextMentionNotificationEvent`
 *
 * The check is made against the event type and event data kind.
 * You should use this guard to safely check the webhook event you received
 * when you're expecting a `TextMentionNotificationEvent`.
 *
 * @param event The webhook event received after calling `webhookHandler.verifyRequest()`.
 * @returns A boolean type predicate.
 */
export function isTextMentionNotificationEvent(
  event: WebhookEvent
): event is TextMentionNotificationEvent {
  return event.type === "notification" && event.data.kind === "textMention";
}

/**
 * Type guard to check if a webhook event is a `CustomNotificationEvent`
 *
 * The check is made against the event type and event data kind.
 * You should use this guard to safely check the webhook event you received
 * when you're expecting a `CustomNotificationEvent`.
 *
 * @param event The webhook event received after calling `webhookHandler.verifyRequest()`.
 * @returns A boolean type predicate.
 */
export function isCustomNotificationEvent(
  event: WebhookEvent
): event is CustomNotificationEvent {
  return event.type === "notification" && isCustomKind(event.data.kind);
}

File: packages/liveblocks-python/liveblocks/webhooks.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from typing import Any, Literal, TypedDict, TypeGuard, Union

_SECRET_PREFIX = "whsec_"
_TOLERANCE_IN_SECONDS = 5 * 60  # 5 minutes

_KNOWN_EVENT_TYPES = frozenset(
    {
        "storageUpdated",
        "userEntered",
        "userLeft",
        "roomCreated",
        "roomDeleted",
        "commentCreated",
        "commentEdited",
        "commentDeleted",
        "commentReactionAdded",
        "commentReactionRemoved",
        "commentMetadataUpdated",
        "threadMetadataUpdated",
        "threadCreated",
        "threadDeleted",
        "ydocUpdated",
        "notification",
        "threadMarkedAsResolved",
        "threadMarkedAsUnresolved",
    }
)

NotificationChannel = Literal["email", "slack", "teams", "webPush"]

__all__ = [
    "WebhookHandler",
    "WebhookEvent",
    "NotificationEvent",
    "NotificationChannel",
    "StorageUpdatedEvent",
    "UserEnteredEvent",
    "UserLeftEvent",
    "RoomCreatedEvent",
    "RoomDeletedEvent",
    "CommentCreatedEvent",
    "CommentEditedEvent",
    "CommentDeletedEvent",
    "CommentReactionAddedEvent",
    "CommentReactionRemovedEvent",
    "CommentMetadataUpdatedEvent",
    "ThreadMetadataUpdatedEvent",
    "ThreadCreatedEvent",
    "ThreadDeletedEvent",
    "ThreadMarkedAsResolvedEvent",
    "ThreadMarkedAsUnresolvedEvent",
    "YDocUpdatedEvent",
    "ThreadNotificationEvent",
    "TextMentionNotificationEvent",
    "CustomNotificationEvent",
    "is_thread_notification_event",
    "is_text_mention_notification_event",
    "is_custom_notification_event",
]

# ---------------------------------------------------------------------------
# Event data TypedDicts
# ---------------------------------------------------------------------------


class _StorageUpdatedData(TypedDict):
    roomId: str
    projectId: str
    updatedAt: str  # ISO 8601


class _UserEnteredData(TypedDict):
    projectId: str
    roomId: str
    connectionId: int
    userId: str | None
    userInfo: dict[str, Any] | None
    enteredAt: str  # ISO 8601
    numActiveUsers: int


class _UserLeftData(TypedDict):
    projectId: str
    roomId: str
    connectionId: int
    userId: str | None
    userInfo: dict[str, Any] | None
    leftAt: str  # ISO 8601
    numActiveUsers: int


class _RoomCreatedData(TypedDict):
    projectId: str
    roomId: str
    createdAt: str  # ISO 8601


class _RoomDeletedData(TypedDict):
    projectId: str
    roomId: str
    deletedAt: str  # ISO 8601


class _CommentCreatedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    commentId: str
    createdAt: str  # ISO 8601
    createdBy: str


class _CommentEditedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    commentId: str
    editedAt: str  # ISO 8601


class _CommentDeletedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    commentId: str
    deletedAt: str  # ISO 8601


class _CommentReactionAddedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    commentId: str
    emoji: str
    addedAt: str  # ISO 8601
    addedBy: str


class _CommentReactionRemovedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    commentId: str
    emoji: str
    removedAt: str  # ISO 8601
    removedBy: str


class _CommentMetadataUpdatedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    commentId: str
    updatedAt: str  # ISO 8601
    updatedBy: str


class _ThreadMetadataUpdatedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    updatedAt: str  # ISO 8601
    updatedBy: str


class _ThreadCreatedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    createdAt: str  # ISO 8601
    createdBy: str


class _ThreadDeletedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    deletedAt: str  # ISO 8601


class _ThreadMarkedAsResolvedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    updatedAt: str  # ISO 8601
    updatedBy: str


class _ThreadMarkedAsUnresolvedData(TypedDict):
    projectId: str
    roomId: str
    threadId: str
    updatedAt: str  # ISO 8601
    updatedBy: str


class _YDocUpdatedData(TypedDict):
    projectId: str
    roomId: str
    updatedAt: str  # ISO 8601


class _ThreadNotificationData(TypedDict):
    channel: NotificationChannel
    kind: Literal["thread"]
    projectId: str
    roomId: str
    userId: str
    threadId: str
    inboxNotificationId: str
    createdAt: str  # ISO 8601
    triggeredAt: str  # ISO 8601


class _TextMentionNotificationData(TypedDict):
    channel: NotificationChannel
    kind: Literal["textMention"]
    projectId: str
    roomId: str
    userId: str
    mentionId: str
    inboxNotificationId: str
    createdAt: str  # ISO 8601
    triggeredAt: str  # ISO 8601


class _CustomNotificationData(TypedDict):
    channel: NotificationChannel
    kind: str  # starts with "$"
    projectId: str
    roomId: str | None
    userId: str
    subjectId: str
    inboxNotificationId: str
    createdAt: str  # ISO 8601
    triggeredAt: str  # ISO 8601


# ---------------------------------------------------------------------------
# Event TypedDicts
# ---------------------------------------------------------------------------


class StorageUpdatedEvent(TypedDict):
    type: Literal["storageUpdated"]
    data: _StorageUpdatedData


class UserEnteredEvent(TypedDict):
    type: Literal["userEntered"]
    data: _UserEnteredData


class UserLeftEvent(TypedDict):
    type: Literal["userLeft"]
    data: _UserLeftData


class RoomCreatedEvent(TypedDict):
    type: Literal["roomCreated"]
    data: _RoomCreatedData


class RoomDeletedEvent(TypedDict):
    type: Literal["roomDeleted"]
    data: _RoomDeletedData


class CommentCreatedEvent(TypedDict):
    type: Literal["commentCreated"]
    data: _CommentCreatedData


class CommentEditedEvent(TypedDict):
    type: Literal["commentEdited"]
    data: _CommentEditedData


class CommentDeletedEvent(TypedDict):
    type: Literal["commentDeleted"]
    data: _CommentDeletedData


class CommentReactionAddedEvent(TypedDict):
    type: Literal["commentReactionAdded"]
    data: _CommentReactionAddedData


class CommentReactionRemovedEvent(TypedDict):
    type: Literal["commentReactionRemoved"]
    data: _CommentReactionRemovedData


class CommentMetadataUpdatedEvent(TypedDict):
    type: Literal["commentMetadataUpdated"]
    data: _CommentMetadataUpdatedData


class ThreadMetadataUpdatedEvent(TypedDict):
    type: Literal["threadMetadataUpdated"]
    data: _ThreadMetadataUpdatedData


class ThreadCreatedEvent(TypedDict):
    type: Literal["threadCreated"]
    data: _ThreadCreatedData


class ThreadDeletedEvent(TypedDict):
    type: Literal["threadDeleted"]
    data: _ThreadDeletedData


class ThreadMarkedAsResolvedEvent(TypedDict):
    type: Literal["threadMarkedAsResolved"]
    data: _ThreadMarkedAsResolvedData


class ThreadMarkedAsUnresolvedEvent(TypedDict):
    type: Literal["threadMarkedAsUnresolved"]
    data: _ThreadMarkedAsUnresolvedData


class YDocUpdatedEvent(TypedDict):
    type: Literal["ydocUpdated"]
    data: _YDocUpdatedData


class ThreadNotificationEvent(TypedDict):
    type: Literal["notification"]
    data: _ThreadNotificationData


class TextMentionNotificationEvent(TypedDict):
    type: Literal["notification"]
    data: _TextMentionNotificationData


class CustomNotificationEvent(TypedDict):
    type: Literal["notification"]
    data: _CustomNotificationData


NotificationEvent = Union[
    ThreadNotificationEvent,
    TextMentionNotificationEvent,
    CustomNotificationEvent,
]

WebhookEvent = Union[
    StorageUpdatedEvent,
    UserEnteredEvent,
    UserLeftEvent,
    RoomCreatedEvent,
    RoomDeletedEvent,
    CommentCreatedEvent,
    CommentEditedEvent,
    CommentDeletedEvent,
    CommentReactionAddedEvent,
    CommentReactionRemovedEvent,
    CommentMetadataUpdatedEvent,
    ThreadMetadataUpdatedEvent,
    ThreadCreatedEvent,
    ThreadDeletedEvent,
    ThreadMarkedAsResolvedEvent,
    ThreadMarkedAsUnresolvedEvent,
    YDocUpdatedEvent,
    ThreadNotificationEvent,
    TextMentionNotificationEvent,
    CustomNotificationEvent,
]


# ---------------------------------------------------------------------------
# Type guards
# ---------------------------------------------------------------------------


def _is_custom_kind(value: object) -> bool:
    return isinstance(value, str) and value.startswith("$")


def is_thread_notification_event(event: WebhookEvent) -> TypeGuard[ThreadNotificationEvent]:
    """Check whether *event* is a thread notification event."""
    return event["type"] == "notification" and event["data"].get("kind") == "thread"


def is_text_mention_notification_event(event: WebhookEvent) -> TypeGuard[TextMentionNotificationEvent]:
    """Check whether *event* is a text-mention notification event."""
    return event["type"] == "notification" and event["data"].get("kind") == "textMention"


def is_custom_notification_event(event: WebhookEvent) -> TypeGuard[CustomNotificationEvent]:
    """Check whether *event* is a custom notification event."""
    return event["type"] == "notification" and _is_custom_kind(event["data"].get("kind"))


# ---------------------------------------------------------------------------
# WebhookHandler
# ---------------------------------------------------------------------------


class WebhookHandler:
    """Verify incoming Liveblocks webhook requests.

    Usage::

        handler = WebhookHandler("whsec_...")
        event = handler.verify_request(headers=request.headers, raw_body=request.body)
    """

    def __init__(self, secret: str) -> None:
        if not secret or not isinstance(secret, str):
            raise ValueError("Secret is required and must be a non-empty string")

        if not secret.startswith(_SECRET_PREFIX):
            raise ValueError("Invalid secret, must start with whsec_")

        secret_key = secret[len(_SECRET_PREFIX) :]
        try:
            self._secret_bytes = base64.b64decode(secret_key)
        except Exception:
            raise ValueError(
                "Webhook secret contains invalid base64 after the 'whsec_' prefix. "
                "Please copy the full secret from your Liveblocks dashboard."
            ) from None

    def verify_request(self, *, headers: dict[str, str], raw_body: str) -> WebhookEvent:
        """Verify a webhook request and return the parsed event.

        Args:
            headers: The HTTP headers as a string-to-string mapping.
            raw_body: The raw request body as a string (do **not** parse it first).

        Returns:
            The parsed webhook event dictionary.

        Raises:
            ValueError: If the request cannot be verified.
        """
        webhook_id, timestamp, raw_signatures = self._verify_headers(headers)

        if not isinstance(raw_body, str):
            raise ValueError(
                f"Invalid raw_body, must be a string, got {type(raw_body).__name__!r} instead. "
                "Make sure you pass the raw request body string, not a parsed object."
            )

        self._verify_timestamp(timestamp)

        signature = self._sign(f"{webhook_id}.{timestamp}.{raw_body}")

        expected_signatures = [
            parts[1] for raw_sig in raw_signatures.split(" ") if len(parts := raw_sig.split(",")) > 1
        ]

        if signature not in expected_signatures:
            raise ValueError(f"Invalid signature, expected one of {', '.join(expected_signatures)}, got {signature}")

        event: WebhookEvent = json.loads(raw_body)

        self._verify_event_type(event)

        return event

    # -- private helpers ----------------------------------------------------

    @staticmethod
    def _verify_headers(headers: dict[str, str]) -> tuple[str, str, str]:
        normalized = {k.lower(): v for k, v in headers.items()}

        webhook_id = normalized.get("webhook-id")
        if not isinstance(webhook_id, str):
            raise ValueError("Invalid webhook-id header")

        timestamp = normalized.get("webhook-timestamp")
        if not isinstance(timestamp, str):
            raise ValueError("Invalid webhook-timestamp header")

        raw_signatures = normalized.get("webhook-signature")
        if not isinstance(raw_signatures, str):
            raise ValueError("Invalid webhook-signature header")

        return webhook_id, timestamp, raw_signatures

    def _sign(self, content: str) -> str:
        mac = hmac.new(self._secret_bytes, content.encode(), hashlib.sha256)
        return base64.b64encode(mac.digest()).decode()

    @staticmethod
    def _verify_timestamp(timestamp_header: str) -> None:
        try:
            timestamp = int(timestamp_header)
        except ValueError:
            raise ValueError("Invalid timestamp") from None

        now = int(time.time())

        if timestamp < now - _TOLERANCE_IN_SECONDS:
            raise ValueError("Timestamp too old")

        if timestamp > now + _TOLERANCE_IN_SECONDS:
            raise ValueError("Timestamp in the future")

    @staticmethod
    def _verify_event_type(event: WebhookEvent) -> None:
        event_type = event.get("type") if isinstance(event, dict) else None

        if event_type and event_type in _KNOWN_EVENT_TYPES:
            if event_type == "notification":
                kind = event["data"].get("kind")
                if kind in ("thread", "textMention") or _is_custom_kind(kind):
                    return
                raise ValueError(f"Unknown notification kind: {kind!r}")
            return

        raise ValueError("Unknown event type, please upgrade to a higher version of liveblocks")
